### PR TITLE
ProofOfPlay should call stream callback on exceptions as well.

### PR DESCRIPTION
**Pull request by hkaya**
This is related to VBID-3732. As it turned out, we need the same thing for proof of plays.

I've also updated the tests. Failed proof of plays should not be kept in internal buffer.

### CODE REVIEW

File: src/proof_of_play.coffee [hunk 1]
```diff
@@ -33,9 +33,15 @@ class ProofOfPlay extends Transform
...
         @_process(response, callback)
+      .catch (err) =>
+        @log.write name: 'ProofOfPlay', message: 'confirm failed', meta: ad
+        callback()
     else
       @expire(ad).then (response) =>
         @_process(response, callback)
+      .catch (err) =>
+        @log.write name: 'ProofOfPlay', message: 'expire failed', meta: ad
+        callback()
```
Comment (markchadwick) [inline src/proof_of_play.coffee]: Could you @write(pop); callback() ?

File: test/proof_of_play_spec.coffee [hunk 2]
```diff
@@ -84,7 +84,7 @@ describe 'ProofOfPlay', ->
       @http.match url: @popUrl, type: 'GET', (req, resolve, reject) =>
         reject()
 
-    it 'should leave the PoP in the internal buffer', (done) ->
+    it 'should drop the PoP request', (done) ->
```
Comment (markchadwick) [inline test/proof_of_play_spec.coffee]: I think I disagree here.

File: src/proof_of_play.coffee [hunk 3]
```diff
@@ -30,12 +30,31 @@ class ProofOfPlay extends Transform
...
       @confirm(ad).then (response) =>
         @_process(response, callback)
+      .catch (e) =>
+        callback()
+        # According to W3 XHR spec, if the state is UNSENT, OPENED or the error
+        # flag is set, status code will be 0. Otherwise, status will be set to
+        # HTTP status code. We need to drop the PoP request on server errors.
+        if e?.currentTarget?.status == 0
+          @log.write name: 'ProofOfPlay', message: 'confirm failed, adding back to the queue.', meta: ad
+          setTimeout(write.bind(@), 5000)
```
Comment (markchadwick) [inline src/proof_of_play.coffee]: Why not fat-arrow on :33 and skip the bind?

Comment (hkaya) [inline src/proof_of_play.coffee]: No particular reason really. Will send out a new change now. Sorry for so many commits!

**Pull request merged**